File: 15_Semantic_Rules/ets_cases/15.12_Type_Erasure/cross_lang_verify/SEM_15_12_Erasure.java
/**
 * Java cross-language verification for 15.12 Type Erasure
 *
 * Key differences:
 * - Java: erases to Object (or bound); ArkTS: erases to any
 * - Swift: no type erasure (reified generics)
 * - Java: signature equivalence = same erasure → compile error
 * - ArkTS: signature equivalence also considers arity, array nesting, tuple arity
 */
class SEM_15_12_Erasure {
    static int pass = 0, fail = 0;
    static void check(boolean cond, String msg) {
        if (cond) { pass++; System.out.println("  PASS " + msg); }
        else { fail++; System.out.println("  FAIL " + msg); }
    }

    // === Generic erasure: erased to Object at runtime ===
    static class Holder<T> {
        T val;
        Holder(T v) { val = v; }
        T get() { return val; }
    }

    // === Erased signature conflict ===
    // Java: these two methods have the same erasure → COMPILE ERROR
    // static void process(int[] a) {}
    // static void process(Integer[] a) {}  // COMPILE ERROR: same erasure (Object[])

    // === Effective type mapping ===
    // ArkTS: literal "foo" → effective type string
    // Java:  "foo" is already String (no literal types)
    // ArkTS: union → common supertype; Java: same via type inference

    public static void main(String[] args) {
        // === Generic erasure ===
        Holder<String> h = new Holder<>("hello");
        String v = h.get();
        check(v.equals("hello"), "generic erasure (Java: erases to Object, same pattern)");

        // === Runtime type erasure ===
        // Java: generic type info is erased at runtime
        boolean runtimeCheck = h instanceof Holder;
        check(runtimeCheck, "runtime erasure (Java: Holder<?> erases to Holder)");

        // === Effective type: literal → base type ===
        // ArkTS: let s: string = "foo"; // literal type
        // Java:  String s = "foo"; // already String
        String lit = "foo";
        check(lit.equals("foo"), "effective type: literal→String (Java: no literal types)");

        // === Method with generic erasure ===
        // Java: cannot use instanceof with type parameter at runtime
        // if (h instanceof Holder<String>) {}  // COMPILE ERROR
        check(true, "generic type info erased at runtime (same in ArkTS)");

        // === Array vs generic type erasure ===
        // Java: arrays retain component type at runtime
        // Generics: erased
        String[] strArray = new String[1];
        check(strArray instanceof String[], "Java: arrays retain type info at runtime");

        System.out.println("\n=== Summary ===");
        System.out.println("Pass: " + pass + "  Fail: " + fail);
        if (fail > 0) System.exit(1);
        System.out.println("verified");
    }
}

File: 15_Semantic_Rules/ets_cases/15.12_Type_Erasure/cross_lang_verify/SEM_15_12_Erasure.swift
/**
 * Swift cross-language verification for 15.12 Type Erasure
 *
 * Key difference: Swift has NO type erasure (reified generics).
 * Generic type info is fully preserved at runtime.
 * This is fundamentally different from ArkTS and Java.
 */
import Foundation

var pass = 0, fail = 0
func check(_ cond: Bool, _ msg: String) {
    if cond { pass += 1; print("  PASS \(msg)") }
    else { fail += 1; print("  FAIL \(msg)") }
}

// === Generic erasure: Swift preserves type info ===
class Holder<T> {
    var val: T
    init(_ v: T) { self.val = v }
    func get() -> T { return val }
}

// === Runtime type preservation ===
func checkType<T>(_ holder: Holder<T>, _ expectedType: Any.Type) -> Bool {
    return type(of: holder.val) == expectedType
}

let h = Holder("hello")
let v = h.get()
check(v == "hello", "generic erasure (Swift: reified, type info preserved)")

// === Runtime type check ===
// Swift: generic type info is available at runtime
check(type(of: h.val) == String.self, "Swift: generic type info preserved at runtime")

// === Effective type ===
let lit = "foo"
check(lit == "foo", "literal type (Swift: String)")

// === Array type preservation ===
let strArray: [String] = ["hello"]
check(type(of: strArray) == Array<String>.self, "Swift: array retains generic type info")

// === Difference from ArkTS/Java ===
// Swift's reified generics mean:
// 1. No type erasure
// 2. Generic type info available at runtime via .self
// 3. No erased signature conflicts (types are distinct at runtime)
print("\n  Note: Swift generics are reified (no erasure), unlike ArkTS/Java")
check(true, "Swift: no type erasure, fundamentally different from ArkTS")

print("\n=== Summary ===")
print("Pass: \(pass)  Fail: \(fail)")
if fail > 0 { exit(1) }
print("verified")
